src/postgres-util/src/privileges.rs

```diff
@@ -0,0 +1,84 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use anyhow::anyhow;
+use postgres_array::{Array, Dimension};
+
+use crate::{sanitize_name_for_catalog_matching, Config, PostgresError};
+
+/// Ensure that the user specified in `config` has `SELECT` privileges for the identified `tables`.
+///
+/// `tables`'s elements should be of the structure `[<schema name>, <table name>]`.
+///
+/// # Panics
+/// If `config` does not specify a user.
+pub async fn check_select_privileges(
+    config: &Config,
+    tables: Vec<[String; 2]>,
+) -> Result<(), PostgresError> {
+    let client = config.connect("check_privileges").await?;
+
+    let tables_len = tables.len();
+
+    let tables = Array::from_parts(
+        tables
+            .into_iter()
+            .map(|i| i.to_vec())
+            .flatten()
+            .map(|i| sanitize_name_for_catalog_matching(&i).to_string())
+            .collect(),
+        vec![
+            Dimension {
+                len: i32::try_from(tables_len).expect("fewer than i32::MAX tables"),
+                lower_bound: 0,
+            },
+            Dimension {
+                len: 2,
+                lower_bound: 0,
+            },
+        ],
+    );
+
+    let invalid_read_privileges = client
+        .query(
+            "
+            SELECT
+                table_catalog || '.' || table_schema || '.' || table_name AS t
```

Comment:
These surely need escaping and quoting? What if a user has a table name with a double quote or a space?

Reply:
```sql
create table """select foo""" (a int);

select distinct table_catalog || '.' || table_schema || '.' || table_name from information_schema.table_privileges where table_name = '"select foo"';
```
```
sean.public."select foo"
```

This generates an error akin to other errors in the process of creating sources, e.g.:

```
CREATE SOURCE "mz_source"
  FROM POSTGRES CONNECTION pg (PUBLICATION 'mz_source')
  FOR TABLES (
    "fodo"
  ) with (size = '1');
ERROR:  table fodo not found in source
```
```
CREATE SOURCE "mz_source"
  FROM POSTGRES CONNECTION pg (PUBLICATION 'mz_source')
  FOR TABLES (
    """fodo"""
  ) with (size = '1');
ERROR:  table "fodo" not found in source
materialize=> 
```

Reply:
Ah, using `format` with `%I` formatting instructions does all of this nonsense more cleanly and for you. TIL.